app/src/main/java/com/kamron/pogoiv/MainActivity.java

```diff
@@ -375,21 +375,28 @@ public void onAppUpdateEvent(AppUpdateEvent event) {
      * Sets up the x,y coordinates of the arc using the trainer level, stores it in Data.arcX/arcY
      */
     private void setupArcPoints() {
-        final int indices = Math.min((int) ((trainerLevel + 1.5) * 2) - 1, 79);
+        /*
+         * Pokemon levels go from 1 to trainerLevel + 1.5, in increments of 0.5.
+         * Here we use dblLevelShifted for levels that are doubled and shifted by - 2; after this adjustment,
+         * the level can be used to index CpM, arcX and arcY.
+         */
+        int maxPokeDblLevelShifted = trainerLevel * 2 + 1; //= (trainerLevel + 1.5) * 2 - 2
+        final int indices = Math.min(maxPokeDblLevelShifted + 1, 79);
         Data.arcX = new int[indices];
         Data.arcY = new int[indices];
 
         double maxAngle = 178.4;
-        double levelCoeff = maxAngle * Data.CpM[trainerLevel * 2 - 2] / (Data.CpM[(int) ((trainerLevel + 1.5) * 2 - 2)] - Data.CpM[0]);
+        double baseCpM = Data.CpM[0];
+        double maxCpmDelta = Data.CpM[maxPokeDblLevelShifted] - baseCpM;
 
-        for (double pokeLevel = 1.0; pokeLevel <= trainerLevel + 1.5; pokeLevel += 0.5) {
-            double angleInDegrees = (Data.CpM[(int) (pokeLevel * 2 - 2)] - Data.CpM[0]) * levelCoeff / Data.CpM[trainerLevel * 2 - 2];
+        for (int pokeDblLevelShifted = 0; pokeDblLevelShifted <= maxPokeDblLevelShifted; pokeDblLevelShifted++) {
```

Comment:
Baaaaaad! You already have the computed size (previous indices) of the Array, so use that. You'll cause an out of bounds exception here if trainer level is 40.